### DISCUSSION TITLE
What's the difference between enforce and condense?

### DISCUSSION THREAD

**Discussion by gatling-nrl** [kinnala/scikit-fem#766] [Q&A] [answered]
Is there a time when I can only use one of them? Is one faster? I'm not sure how to chose between them.

**Comment by gdmcbain** [comment, marked as answer]
Yeah, for a long time we only had `condense`.  I was pretty happy with it, although I did (& still do) plan to try out different techniques of imposing Dirichlet/essential boundary conditions in #753 .

`enforce` was introduced following #591. There's a long discussion there.  Quite a productive discussion, as I recall.

Basically condense eliminates the degrees of freedom constrained by essential boundary conditions; enforce leaves them in but modifies the system of equations so that in the solution they have to have the required values.

Another alternative is `penalize`, as in 

https://github.com/kinnala/scikit-fem/blob/7500d1ca3523196b8d88e6c2cbf7cfecd6c52331/docs/examples/ex19.py#L67

`enforce` is convenient for some cases as explained in #591, but it does have the disadvantage that it destroys symmetry, in case you're using a solver that relies on that.  (Although actually I believe it is becoming more common to use Krylov solvers that don't rely on symmetry even when the problem is nominally symmetric; e.g. GMRES instead of CG or MINRES.

For direct solvers, I wouldn't think that there would be much difference in speed between any of these three, but I haven't timed them.  It's usually more about what's most convenient and tidier in the scripts and this will vary depending on the example (steady or unsteady, linear or nonlinear, &c.).

For a good discussion in the literature of different ways of handling essential conditions in the finite element method, I recommend Ern & Guermond, as cited [2019](https://github.com/kinnala/scikit-fem/issues/150#issuecomment-480118703) in #150.
